src/Compilers/CSharp/Test/Semantic/Semantics/NativeIntTests.cs

```diff
@@ -40,10 +42,232 @@ public void LanguageVersion()
             comp.VerifyDiagnostics();
         }
 
+        /// <summary>
+        /// System.IntPtr and System.UIntPtr definitions from metadata.
+        /// </summary>
+        [Fact]
+        public void TypeDefinitions_FromMetadata()
+        {
+            var source =
+@"interface I
+{
+    void F1(System.IntPtr x, nint y);
+    void F2(System.UIntPtr x, nuint y);
+}";
+            var comp = CreateCompilation(source, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyDiagnostics();
+
+            var method = comp.GetMember<MethodSymbol>("I.F1");
+            Assert.Equal("void I.F1(System.IntPtr x, nint y)", method.ToDisplayString(FormatWithSpecialTypes));
+            VerifyTypes((NamedTypeSymbol)method.Parameters[0].Type, (NamedTypeSymbol)method.Parameters[1].Type, signed: true);
+
+            method = comp.GetMember<MethodSymbol>("I.F2");
+            Assert.Equal("void I.F2(System.UIntPtr x, nuint y)", method.ToDisplayString(FormatWithSpecialTypes));
+            VerifyTypes((NamedTypeSymbol)method.Parameters[0].Type, (NamedTypeSymbol)method.Parameters[1].Type, signed: false);
+        }
+
+        /// <summary>
+        /// System.IntPtr and System.UIntPtr definitions from source.
+        /// </summary>
+        [Fact]
+        public void TypeDefinitions_FromSource()
+        {
+            var sourceA =
+@"namespace System
+{
+    public class Object { }
+    public abstract class ValueType { }
+    public struct Void { }
+    public struct IntPtr { }
```

Comment:
Are these types (`IntPtr`, `UIntPtr`)  native ints?

Reply:
Type definitions return `false` for `IsNativeInt`. The only `NamedTypeSymbol` instances that should return `true` for `IsNativeInt` are type references of the form `nint` or `nuint` (from source) or type references to `System.[U]IntPtr` with `[NullableInteger]` (from metadata).

Added checks for `IsNativeInt` for these type definitions.

---
In reply to: [375001603](https://github.com/dotnet/roslyn/pull/40994#discussion_r375001603) [](ancestors = 375001603)